reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -5,311 +5,236 @@
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 import java.util.concurrent.atomic.AtomicLong;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
-import java.util.function.LongSupplier;
-import java.util.function.Supplier;
 import java.util.stream.Stream;
 
 import reactor.core.Disposable;
 import reactor.core.Disposables;
 import reactor.core.Exceptions;
 import reactor.core.Scannable;
-import reactor.util.annotation.Nullable;
 
 /**
- * Dynamically creates ScheduledExecutorService-based Workers and caches the thread pools, reusing
- * them once the Workers have been shut down. This scheduler is time-capable (can schedule
- * with delay / periodically).
- * <p>
- * The maximum number of created thread pools is capped. Tasks submitted after the cap has been
- * reached can be enqueued up to a second limit, possibly lifted by using {@link Integer#MAX_VALUE}.
- * <p>
- * The default time-to-live for unused thread pools is 60 seconds, use the
- * appropriate constructor to set a different value.
- * <p>
- * This scheduler is not restartable.
+ * Scheduler that hosts a pool of 0-N single-threaded {@link BoundedScheduledExecutorService} and exposes workers
+ * backed by these executors, making it suited for moderate amount of blocking work. Note that requests for workers
+ * will pick an executor in a round-robin fashion, so tasks from a given worker might arbitrarily be impeded by
+ * long-running tasks of a sibling worker (and tasks are pinned to a given executor, so they won't be stolen
+ * by an idle executor).
+ *
+ * This scheduler is time-capable (can schedule with delay / periodically).
  *
  * @author Simon Baslé
  */
-final class BoundedElasticScheduler
-		implements Scheduler, Supplier<ScheduledExecutorService>, Scannable {
+final class BoundedElasticScheduler implements Scheduler, Scannable {
 
-	static final AtomicLong COUNTER = new AtomicLong();
+	static final int DEFAULT_TTL_SECONDS = 60;
+
+	static final AtomicLong EVICTOR_COUNTER = new AtomicLong();
 
 	static final ThreadFactory EVICTOR_FACTORY = r -> {
-		Thread t = new Thread(r, "elasticBounded-evictor-" + COUNTER.incrementAndGet());
+		Thread t = new Thread(r, Schedulers.BOUNDED_ELASTIC + "-evictor-" + EVICTOR_COUNTER.incrementAndGet());
 		t.setDaemon(true);
 		return t;
 	};
 
-	static final CachedService SHUTDOWN            = new CachedService(null);
-	static final int           DEFAULT_TTL_SECONDS = 60;
-
-	final ThreadFactory              factory;
-	final int                        ttlSeconds;
-	final int                        threadCap;
-	final int                        deferredTaskCap;
-	final Deque<CachedServiceExpiry> idleServicesWithExpiry;
-	final Queue<DeferredFacade>      deferredFacades;
-	final Queue<CachedService>       allServices;
-	final ScheduledExecutorService   evictor;
+	static final int             DISPOSED = -1;
```

Comment:
This constant seems to be reused between `BoundedServices` and `BoundedState`.

I would suggest having two, in each class, especially given that `BoundedServices` uses it for checking a variable that is a counter (although I haven't checked `BoundedState`), so that the value is more meaningless in the given domain

Reply:
they both are atomic counters: `BoundedServices` counts the number of states, aka executors it spawned. `BoundedState` counts the number of `Worker`s that picked that state/executor.

Reply:
I'll duplicate the constant and rename the second one `EVICTED`